### PR TITLE
Attempt to pipe logit_bias to sampler's embedding_bias

### LINKED ISSUE #379: [Feature] Add support for `logit_bias`
Support just landed in `transformers` lib for this. See:
- [`SequenceBiasLogitsProcessor`](https://huggingface.co/docs/transformers/main/en/internal/generation_utils#transformers.SequenceBiasLogitsProcessor)
- [Corresponding `GenerationConfig`](https://huggingface.co/docs/transformers/main/en/main_classes/text_generation#transformers.GenerationConfig.sequence_bias)
- [discussion](https://github.com/huggingface/transformers/issues/22168#issuecomment-1477998997)

### PR DESCRIPTION
**Pull request by benbot**
`logit_bias` is an important feature of the OpenAI API that vllm seems to have implemented but not exposed to the actual api.

This is me taking a crack on exposing that functionality.

For the life of me, I can't get my cuda versions to all agree to build this locally, so while I try to do that i'm opening the PR for others to try out.

Should resolve: #379

### CODE REVIEW

**Comment by viktor-ferenczi** [comment]
Please rebase this branch to current main, then I will build and test this. I also want to get LMQL working with vLLM, because performance is not good with any of the other backends.

File: vllm/sampling_params.py [hunk 1]
```diff
@@ -92,6 +93,7 @@ def __init__(
         stop_token_ids: Optional[List[int]] = None,
         ignore_eos: bool = False,
         max_tokens: int = 16,
+        logit_bias: float = [],
```
Comment (viktor-ferenczi) [inline vllm/sampling_params.py:96]: Proper type is: `Optional[List[float]]`

File: vllm/sampling_params.py [hunk 2]
```diff
@@ -122,6 +124,7 @@ def __init__(
         self.max_tokens = max_tokens
         self.logprobs = logprobs
         self.prompt_logprobs = prompt_logprobs
+        self.logit_bias = logit_bias
```
Comment (viktor-ferenczi) [inline vllm/sampling_params.py:127]: Define type here as well

Comment (viktor-ferenczi) [reply vllm/sampling_params.py:127]: Also add `self.logit_bias` to `__repr__` below.

**Review by viktor-ferenczi — CHANGES_REQUESTED** [review]
Looks good, added minor comments.

The client must know the vocabulary and the vocab_size in order to pass a `logits_bias` array which works with the model loaded. Right now the client has to load the tokenizer corresponding to the model loaded into the vLLM server to achieve this. It makes clients more complex and there is a risk of using the wrong tokenizer for the model, ending up in errors.

I suggest exposing two new REST API calls:
- `vocabulary` Returns the vocabulary (array of strings). The `vocab_size` is the length of this array. It may also return the special tokens, stop tokens, start/end of conversation tokens, etc.
- `tokenize` Accepts `text` and returns an array of integers (token values).

This way no tokenizer needs to be built on client side and no way to use the wrong vocabulary. The client can build the `logits_bias` based on the information returned (the client must acquire the vocabulary once on initialization).

Having these would allow for a pretty straightforward test case and easier integration with LMQL.

File: vllm/model_executor/layers/sampler.py [hunk 3]
```diff
@@ -217,6 +217,11 @@ def _apply_penalties(
     logits -= presence_penalties.unsqueeze(dim=1) * mask
     return logits
 
+def _get_logit_bias(input_metadata: InputMetadata) -> any:
+    logit_biases: any = []
+    for seq_group in input_metadata.seq_groups:
+        set_ids, sampling_params = seq_group
+        logit_biases += [sampling_params.logit_bias]
```
Comment (viktor-ferenczi) [inline vllm/model_executor/layers/sampler.py:224]: Add explicit validation of the size of the `logit_bias` array received from `sampling_params`. Having an explicit error with a clear explanation here is better than ending up with a cryptic PyTorch error message while trying to add arrays of mismatching sizes later.

(It cannot be validated inside `SamplingParams`, because the right size is not known there.)

**Review by viktor-ferenczi — CHANGES_REQUESTED** [review]

**Comment by viktor-ferenczi** [comment]
Related to #535

**Comment by viktor-ferenczi** [comment]
> > The client must know the vocabulary and the vocab_size in order to pass a `logits_bias` array which works with the model loaded.
> 
> @viktor-ferenczi I may be missing something, but why do they need to know the vocab size? Isn't logit_bias a sparse mapping of token ids to bias to apply?

I did not know it is a sparse mapping, I expected it to be an array of floats with `vocab_size` items. 

Anyway, the client must be able to retrieve the vocabulary of the currently loaded model for the solution to be fully usable. Otherwise the client must instantiate the tokenizer of the exact same model loaded into the vLLM server just to get the vocabulary, which means a lot more dependencies and room for error, not to mention the added loading time.

**Comment by creatorrr** [comment]
@viktor-ferenczi @benbot did you guys come to a consensus on how to go about implementing this? Loading the tokenizer vocab will be tricky but then again this feature is meant for rather advanced use cases and we could just leave it to the api consumer to figure out the mapping between token_ids and tokens (that's what openai did anyway).

**Comment by viktor-ferenczi** [comment]
There is a `logits_processors` parameter in `SamplingParams` already. It was added by #1469 which has been merged to `main` already.

**Comment by viktor-ferenczi** [comment]
Depending on your use case using a grammar would be a much cleaner alternative than trying to manipulate logits, it also moves the problem of handling tokens into the server completely. 

See #2105: Add grammars
